src/SqliteContractCache.ts

```diff
@@ -0,0 +1,245 @@
+  private readonly logger = LoggerFactory.INST.create("SQLite contract cache");
+
+  private _db: Database;
+
+  // Lazy initialization upon first access
+  private get db() {
+    if (!this._db) {
+      if (this.cacheOptions.inMemory) {
+        this._db = new Database(":memory:");
+      } else {
+        if (!this.cacheOptions.dbLocation) {
+          throw new Error(
+            "LevelDb cache configuration error - no db location specified"
+          );
+        }
+        const dbLocation = this.cacheOptions.dbLocation;
+        this.logger.info(`Using location ${dbLocation}`);
+        this._db = new Database(dbLocation + ".db");
+      }
+
+      this._db.pragma("journal_mode = WAL");
+      this.sortKeyTable();
+    }
+    return this._db;
+  }
+
+  private sortKeyTable() {
+    this.db.exec(`
+        CREATE TABLE IF NOT EXISTS sort_key_cache
+        (
+            id
+            INTEGER
+            PRIMARY
+            KEY,
+            key
+            TEXT,
+            sort_key
+            TEXT,
+            value
+            jsonb,
+            UNIQUE
+        (
+            key,
+            sort_key
+        )
+            )
+    `);
+  }
+
+  constructor(
+    private readonly cacheOptions: CacheOptions,
+    private readonly sqliteCacheOptions?: SqliteCacheOptions
+  ) {
+    if (!sqliteCacheOptions) {
+      this.sqliteCacheOptions = {
+        maxEntriesPerContract: 10,
+      };
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async get(
+    cacheKey: CacheKey,
+    returnDeepCopy?: boolean
+  ): Promise<SortKeyCacheResult<V> | null> {
+    const result = this.db
+      .prepare(
+        `SELECT value
+         FROM sort_key_cache
+         WHERE key = ?
+           AND sort_key = ?;`
+      )
+      .pluck()
+      .get(cacheKey.key, cacheKey.sortKey);
+
+    if (result) {
+      return new SortKeyCacheResult<V>(cacheKey.sortKey, JSON.parse(result));
+    }
+    return null;
+  }
+
+  async getLast(key: string): Promise<SortKeyCacheResult<V> | null> {
+    const result = this.db
+      .prepare(
+        "SELECT sort_key, value FROM sort_key_cache WHERE key = ? ORDER BY sort_key DESC LIMIT 1"
+      )
+      .get(key);
+
+    if (result && result.value) {
+      return new SortKeyCacheResult<V>(
+        result.sort_key,
+        JSON.parse(result.value)
+      );
+    }
+    return null;
+  }
+
+  async getLessOrEqual(
+    key: string,
+    sortKey: string
+  ): Promise<SortKeyCacheResult<V> | null> {
+    const result = this.db
+      .prepare(
+        "SELECT sort_key, value FROM sort_key_cache WHERE key = ? AND sort_key <= ? ORDER BY sort_key DESC LIMIT 1"
+      )
+      .get(key, sortKey);
+
+    if (result && result.value) {
+      return new SortKeyCacheResult<V>(
+        result.sort_key,
+        JSON.parse(result.value)
+      );
+    }
+    return null;
+  }
+
+  async put(stateCacheKey: CacheKey, value: V): Promise<void> {
+    const strVal = JSON.stringify(value);
```

Comment:
maybe we should use safe-stable-strinigify here, to get stable and detereministic results?

Reply:
yes, definitely, changed